Update native routing CIDR flags description

jibi opened the pull request:
See individual commits

jibi committed:
daemon: update ipv{4,6}-native-routing-cidr flag description

Suggested-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

kkourt commented on daemon/cmd/daemon_main.go:
```diff
@@ -648,12 +648,19 @@ func initializeFlags() {
 	flags.String(option.NativeRoutingCIDR, "",
-		fmt.Sprintf("Allows to explicitly specify the IPv4 CIDR for native routing. This value corresponds to the configured cluster-cidr. Deprecated in favor of --%s", option.IPv4NativeRoutingCIDR))
+		fmt.Sprintf("Allows to explicitly specify the IPv4 CIDR for native routing."+
+			"When specified, Cilium assumes something else has configured the networking for this range and will hand traffic destined to that range to the Linux network stack without applying any SNAT. "+
```
```suggestion
			"When specified, Cilium assumes networking for this CIDR is preconfigured and it will hand traffic destined to that range to the Linux network stack without applying any SNAT. "+
```

joestringer replied:
One addendum to @kkourt's suggestion above, wherever possible try to avoid "it" as this can often add ambiguity to statements. `it` -> `Cilium` will clarify further.

kkourt replied:
@joestringer isn't two consecutive Cilium's too much?

Maybe:
"When specified, Cilium assumes networking for this CIDR is preconfigured and hands traffic destined for that range to the Linux network stack without applying any SNAT. "

joestringer replied:
yeah I thought about that as well, but your latest proposal seems even better.

kkourt reviewed (approved):
Thanks Gilberto, LGTM!

(just a minor wording suggestion)

jibi committed:
install: update ipv{4,6}NativeRoutingCIDR flag description

Suggested-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

kkourt reviewed (changes requested):
Thanks Gilberto!

How about adding an appropriate commit message on the third commit that explains the reasoning behind this?

Here's a suggestion for the message, based on what @joestringer said during our offline discussion (slightly modified):

```
Issue a warning when these two options are selected, since we are not aware of
a use-case where this combination is deliberately chosen.

The reasoning is as follows. If you need auto-direct-node-routes, then all
nodes are directly connected via L2,  you're using direct routing, and you
don't otherwise have a way to share the routes for pods. Hence, what you,
presumably, want is to automatically configure the routes for pod CIDRs between
nodes and also you would not want to masquerade/NAT such intra-cluster traffic.

In fact, if you did apply NAT for such traffic then it could even get you into
trouble because the identity of traffic from pods on one node towards pods on
another node could get lost. This is because in direct routing mode the
identity of the source of the traffic is not held in the packets between nodes,
therefore the destination must use the source address to look up the identity
when applying ingress policy. If the source masqueraded the traffic, then
instead of seeing a source pod IP, you'll see a source node IP and therefore
the identity will be considered as "remote-node" instead of having a rich
identity of the actual pod source, so the network policy will not be able to
allow traffic based on the source pod labels.

```

joestringer commented on pkg/option/config.go:
```diff
@@ -2690,6 +2690,10 @@ func (c *DaemonConfig) Populate() {
 		}
 	}
 
+	if nativeRoutingCIDR == "" && ipv4NativeRoutingCIDR == "" && c.EnableAutoDirectRouting {
```
This should probably be wrapped also in a check against `c.EnableIPv4`. Similar for IPv6 case below. Doesn't make sense to warn users if that support is disabled.

joestringer commented on pkg/option/config.go:
```diff
@@ -2690,6 +2690,10 @@ func (c *DaemonConfig) Populate() {
 	}
 
+	if nativeRoutingCIDR == "" && ipv4NativeRoutingCIDR == "" && c.EnableAutoDirectRouting {
+		log.Warnf("%s will be ineffective since %s is not set", EnableAutoDirectRoutingName, IPv4NativeRoutingCIDR)
```
I don't think that's actually what we're saying. `--auto-direct-node-routes` will still be effective, that logic only creates routes on the nodes based on k8s pod CIDR ranges for each node.

The concern is that traffic between pods on different nodes could be unnecessarily NAT'd, which can cause problems with performance, observability and policy.

brb commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -168,15 +168,15 @@ cilium-agent [flags]
       --iptables-lock-timeout duration                       Time to pass to each iptables invocation to wait for xtables lock acquisition (default 5s)
       --iptables-random-fully                                Set iptables flag random-fully on masquerading rules
-      --ipv4-native-routing-cidr string                      Allows to explicitly specify the IPv4 CIDR for native routing. This value corresponds to the configured cluster-cidr.
+      --ipv4-native-routing-cidr string                      Allows to explicitly specify the IPv4 CIDR for native routing. When specified, Cilium assumes networking for this CIDR is preconfigured and hands traffic destined for that range to the Linux network stack without applying any SNAT. The network stack will then be responsible for routing it to its destination, potentially with any NAT app ...
```
> the user must configure the routes to reach pods, either manually or by setting the auto-direct-node-routes flag

I think it's still not 100%, as in the tunneling mode you don't need to manually configure routes to reach pods. Is ipv4-native-routing used in the tunneling mode?

joestringer replied:
As far as I understand, typically if you're running tunneling mode, you wouldn't bother with ipv4-native-routing. Anything inside the cluster would be handled via tunnels, anything outside the cluster you would masquerade. Though in theory I guess you could enable both at the same time, and as long as you handle the appropriate routing to any destinations within the native routing CIDR range, it could work. Agree though, you wouldn't have to worry about pod ranges as Cilium will handle routing for this traffic.

brb replied:
I think (haven't validated yet) that in the tunneling mode setting the ipv4-native-routing option can have the counterproductive effect. According to my old comment (https://github.com/cilium/cilium/pull/14611#issuecomment-760766052), in the tunnel mode pod2host doesn't go over the tunnel. So, setting the option would avoid SNAT-ing which could result in the packet (pod2host) drop if an underlying network drops packets from unknown subnets (the AWS EC2 default mode).

joestringer replied:
:+1: I could see how that could be a problem yeah.

There's a problem with the other way though, if sending traffic from pod2host and the host has host firewall enabled, then if you NAT the traffic, the destination will see host2host traffic and the policy may allow pod2host because it thinks the traffic is actually host2host. Probably we need to take a closer look at the asymmetric tunnel forwarding issue and decide the right way to handle it for that case.

kkourt replied:
How about, admitting that the semantics of above flag are not really clear, and short term use something like the following for the flag description:

> When specified, Cilium assumes networking for this CIDR is preconfigured and
> hands traffic destined for that range to the Linux network stack without
> applying any SNAT. Generally speaking, specifying a native routing CIDR
> implies that Cilium can depend on the underlying networking stack to route
> packets to their destination. To offer a concrete example, if Cilium is
> configured to use direct routing and the Kubernetes CIDR is included in the
> native routing CIDR, the user must configure the routes to reach pods, either
> manually or by setting the auto-direct-node-routes flag.

ldelossa reviewed (approved):
This LGTM and is a welcome addition of visibility to flag meanings. 

I'll approve contingent to Joe's and Martynas's comments which I also agree with.

brb reviewed (approved):
Thanks!

joestringer commented on pkg/option/config.go:
```diff
@@ -2712,6 +2717,11 @@ func (c *DaemonConfig) Populate() {
 
+	if c.EnableIPv6 && ipv6NativeRoutingCIDR == "" && c.EnableAutoDirectRouting {
+		log.Warnf("%s should not be enabled when %s is not set as this configuration may led to pod to pod traffic being masqueraded, "+
+			"which can cause problems with performance, observability and policy", EnableAutoDirectRoutingName, IPv6NativeRoutingCIDR)
```
I think this is accurate, but can probably be made simpler to understand by proposing the "correct" configuration first rather than just describing what is bad. Proposal:

```suggestion
		log.Warnf("If %s is enabled, then you are recommended to also configure %s. If %s is not configured, this may lead to pod to pod traffic being masqueraded, "+
			"which can cause problems with performance, observability and policy", EnableAutoDirectRoutingName, IPv6NativeRoutingCIDR, IPv6NativeRoutingCIDR)
```

(Same for others if you think this improves readability).

joestringer reviewed (commented):
Just one more nit, but otherwise I think the wording is good :+1:

jibi committed:
daemon: warn if auto direct node routes is set without a native routing CIDR

Issue a warning when these two options are selected, since we are not aware of
a use-case where this combination is deliberately chosen.

The reasoning is as follows. If you need auto-direct-node-routes, then all
nodes are directly connected via L2, you're using direct routing, and you don't
otherwise have a way to share the routes for pods. Hence, what you, presumably,
want is to automatically configure the routes for pod CIDRs between nodes and
also you would not want to masquerade/NAT such intra-cluster traffic.

In fact, if you did apply NAT for such traffic then it could even get you into
trouble because the identity of traffic from pods on one node towards pods on
another node could get lost. This is because in direct routing mode the
identity of the source of the traffic is not held in the packets between nodes,
therefore the destination must use the source address to look up the identity
when applying ingress policy. If the source masqueraded the traffic, then
instead of seeing a source pod IP, you'll see a source node IP and therefore
the identity will be considered as "remote-node" instead of having a rich
identity of the actual pod source, so the network policy will not be able to
allow traffic based on the source pod labels.

Suggested-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
